### PR TITLE
Record: 11L Full Stack + XSA4 + Tight SWA + Late QAT (val_bpb=1.1320)

### PR DESCRIPTION
**Pull request by joelnishanth**
## Record: 11L Full Stack + XSA4 + Tight SWA + Late QAT

**val_bpb: 1.1320** (sliding window, stride=64) | Artifact: 15,753,020 bytes | 8xH100 SXM

## Architecture
- 11 transformer layers, 512-dim, 8 heads (4 KV heads, GQA)
- 3x MLP expansion with relu-squared activation
- XSA on last 4 layers (GQA-aware, zero-alloc)
- Partial RoPE (16/64 dims) + NTK-aware scaling
- LN Scale Factor 1/sqrt(layer_idx+1)
- U-Net skip connections (5 encoder, 6 decoder)
- SmearGate + BigramHash (2048 buckets, dim=128)
- Shared Value Embedding (dim=128, layers 9,10)
- FlashAttention 3 (with FA2 fallback)
- Orthogonal init with proj scaling
- Logit softcap 30.0, tied embeddings

## Training
- Muon: lr=0.025, momentum=0.99, WD=0.04
- AdamW embeddings: lr=0.035, scalars: lr=0.025, WD=0.04
- Gradient clip: 0.3
- Batch: 786,432 tokens/step, seq_len=2048
- Warmdown: 3000 iters (wallclock-based)
- Tight SWA: every 50 steps when scale<0.2 (12 checkpoints)
- Late QAT: STE int6 when LR scale<0.1

## Quantization
- Int6 per-row for MLP + attention weights
- Int8 per-row for embeddings
- zstd level 22 compression

## Results
- 6,374 steps in 600s at 94ms/step
- Pre-quant val_bpb: 1.1475
- Post-SWA val_bpb: 1.1475 (zero SWA penalty)
- Quant gap: ~0.008
- **Sliding window val_bpb: 1.1320**
- Artifact size: 15,753,020 bytes

## Run
```bash
DATA_PATH=../../../data/datasets/fineweb10B_sp1024/ TOKENIZER_PATH=../../../data/tokenizers/fineweb_1024_bpe.model VOCAB_SIZE=1024 torchrun --standalone --nproc_per_node=8 train_gpt.py
```

— [Joel Nishanth](https://offlyn.ai) · [offlyn.AI](https://offlyn.ai)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: 11L Full Stack + XSA4 + Tight SWA + Late QAT (val_bpb=1.1320)

**BPB:** 1.1320 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `f8285185dc20`, file `records/track_10min_16mb/2026-03-20_XSA_QAT_FullStack/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 2.02s, dim=512, layers=11, vocab=1024, code=73064 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 2.02s, dim=512, layers=11, vocab=1024, code=73064 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
